core/server/proxy/src/main/java/alluxio/web/ProxyWebServer.java

```diff
@@ -104,21 +128,51 @@ public void init() throws ServletException {
 
       @Override
       public void service(final ServletRequest req, final ServletResponse res)
-              throws ServletException, IOException {
+          throws ServletException, IOException {
         Stopwatch stopWatch = Stopwatch.createStarted();
         super.service(req, res);
         if ((req instanceof HttpServletRequest) && (res instanceof HttpServletResponse)) {
           HttpServletRequest httpReq = (HttpServletRequest) req;
           HttpServletResponse httpRes = (HttpServletResponse) res;
-          logAccess(httpReq, httpRes, stopWatch);
+          logAccess(httpReq, httpRes, stopWatch, null);
         }
       }
     };
-    ServletHolder servletHolder = new ServletHolder("Alluxio Proxy Web Service", servlet);
-    mServletContextHandler
-        .addServlet(servletHolder, PathUtils.concatPath(Constants.REST_API_PREFIX, "*"));
-    // TODO(czhu): Move S3 API logging out of CompleteMultipartUploadHandler into a logging handler
-    addHandler(new CompleteMultipartUploadHandler(mFileSystem, Constants.REST_API_PREFIX));
+
+    if (Configuration.getBoolean(PropertyKey.PROXY_S3_V2_VERSION_ENABLED)) {
+      super.getServerConnector().addBean(new ProxyListener());
+      ServletHolder s3ServletHolder = new ServletHolder("Alluxio Proxy V2 S3 Service",
+          new S3RequestServlet() {
+            @Override
+            public void init() throws ServletException {
+              super.init();
+              getServletContext().setAttribute(ALLUXIO_PROXY_SERVLET_RESOURCE_KEY, proxyProcess);
+              getServletContext()
+                  .setAttribute(FILE_SYSTEM_SERVLET_RESOURCE_KEY, mFileSystem);
+              getServletContext().setAttribute(STREAM_CACHE_SERVLET_RESOURCE_KEY,
+                  new StreamCache(Configuration.getMs(PropertyKey.PROXY_STREAM_CACHE_TIMEOUT_MS)));
+              getServletContext().setAttribute(ALLUXIO_PROXY_AUDIT_LOG_WRITER_KEY,
+                  mAsyncAuditLogWriter);
+
+              getServletContext().setAttribute(PROXY_S3_V2_LIGHT_POOL,
+                  new ThreadPoolExecutor(8, 64, 0,
+                  TimeUnit.SECONDS, new ArrayBlockingQueue<>(64 * 1024),
+                  ThreadFactoryUtils.build("S3-LIGHTPOOL-%d", false)));
+              getServletContext().setAttribute(PROXY_S3_V2_HEAVY_POOL,
+                  new ThreadPoolExecutor(8, 64, 0,
+                  TimeUnit.SECONDS, new ArrayBlockingQueue<>(64 * 1024),
+                  ThreadFactoryUtils.build("S3-HEAVYPOOL-%d", false)));
+              getServletContext().setAttribute(PROXY_S3_HANDLER_MAP, mS3HandlerMap);
+            }
+          });
+      mServletContextHandler
+          .addServlet(s3ServletHolder, PathUtils.concatPath(Constants.REST_API_PREFIX, "*"));
+    } else {
```

Comment:
can a return be used directly to avoid an else?

Reply:
done